timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/LockWatchEventIntegrationTest.java

```diff
@@ -555,6 +604,123 @@ public Set<LockDescriptor> invalidateSome(Set<LockDescriptor> invalidatedLocks)
+                                random)
+                                .entrySet()
+                                .stream())
+                // Ensure unique rows if metadata is present to prevent commit from failing due to SERIALIZABLE conflict
+                // handler
+                .filterEntries((cell, optMetadata) ->
+                        optMetadata.isEmpty() || seenRowsForMetadata.add(Bytes.from(cell.getRowName())))
+                .collectToMap();
+        // For each cell, 50/50 split on whether it should have metadata
+        return ImmutableMap.of(
+                TABLE_REF,
+                maybeAssociateMetadata(tableOneCells, random),
+                TABLE_2_REF,
+                maybeAssociateMetadata(tableTwoCells, random),
+                SERIALIZABLE_WATCHED_TABLE_REF,
+                serializableTableCellsUniqueRows);
+    }
+
+    private static Map<Cell, Optional<ChangeMetadata>> maybeAssociateMetadata(Set<Cell> cells, Random random) {
+        return KeyedStream.of(cells.stream())
+                .<Optional<ChangeMetadata>>map(_cell ->
+                        random.nextBoolean() ? Optional.of(generateRandomChangeMetadata(random)) : Optional.empty())
+                .collectToMap();
+    }
+
+    private static ChangeMetadata generateRandomChangeMetadata(Random random) {
+        int index = random.nextInt(4);
+        switch (index) {
+            case 0:
+                return ChangeMetadata.created(PtBytes.toBytes(random.nextLong()));
+            case 1:
+                return ChangeMetadata.unchanged();
+            case 2:
+                return ChangeMetadata.updated(PtBytes.toBytes(random.nextLong()), PtBytes.toBytes(random.nextLong()));
+            case 3:
+                return ChangeMetadata.deleted(PtBytes.toBytes(random.nextLong()));
+        }
+        throw new RuntimeException("unreachable");
+    }
+
+    private static Optional<LockRequestMetadata> getLockRequestMetadataForValues(
+            Map<TableReference, Map<Cell, Optional<ChangeMetadata>>> values) {
+        Map<LockDescriptor, ChangeMetadata> locks = KeyedStream.ofEntries(values.entrySet().stream())
+                .filterKeys(LockWatchEventIntegrationTest::isWatchedTable)
+                .flatMapEntries((table, cellToMetadata) -> KeyedStream.ofEntries(cellToMetadata.entrySet().stream())
+                        .flatMap(Optional::stream)
+                        .mapKeys(cell -> getDescriptor(table, cell))
+                        .entries())
+                .collectToMap();
+        return locks.isEmpty() ? Optional.empty() : Optional.of(LockRequestMetadata.of(locks));
+    }
+
+    private static LockDescriptor getDescriptor(TableReference tableRef, Cell cell) {
+        // SERIALIZABLE conflict handler uses row-level locks only
+        if (tableRef.equals(SERIALIZABLE_WATCHED_TABLE_REF)) {
+            return AtlasRowLockDescriptor.of(tableRef.getQualifiedName(), cell.getRowName());
+        }
```

Comment:
I don't like this hard-coded condition, but exporting the conflict handlers from `LockWatchIntegrationTestUtilities` also seemed ugly.